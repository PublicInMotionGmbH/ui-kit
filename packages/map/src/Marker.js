import React from 'react'
import PropTypes from 'prop-types'
import { Marker as GoogleMarker, InfoWindow } from 'react-google-maps'

import { Location } from './PropTypes'

const propTypes = {
  /** Marker position */
  position: Location.isRequired,

  /** Marker information window */
  info: PropTypes.node,

  /** Action on click on marker */
  onClick: PropTypes.func,

  /** Action on closing info window through X button */
  onClose: PropTypes.func,

  /** Is information window opened? */
  open: PropTypes.bool
}

/**
 * Component which represents Marker on map.
 *
 * @property {object} props
 * @property {object|{ lat: number, lng: number}} props.position
 * @property {*} [props.info]
 * @property {boolean} [props.open]
 * @property {function} [props.onClick]
 * @property {function} [props.onClose]
 *
 * @property {object} state
 * @property {boolean} state.open
 *
 * @class
 */
class Marker extends React.PureComponent {
  state = {
    open: this.props.open || false
  }

  /**
   * Handle change between controlled and not controlled component
   *
   * @param {object} props
   * @param {object|{ lat: number, lng: number}} props.position
   * @param {*} [props.info]
   * @param {boolean} [props.open]
   * @param {function} [props.onClick]
   * @param {function} [props.onClose]
   * @param {object} state
   *
   * @returns {object || null}
   */
  static getDerivedStateFromProps (props, state) {
    let nextOpen = null

    // When component is switching to controlled, we keep current open status
    if (state.open != null && props.open == null) {
      nextOpen = state.open
    }

    // When component doesn't have info window, do not open it anytime
    if (props.info == null) {
      nextOpen = false
    }

    // When opening status has changed, update it
    if (nextOpen !== null && nextOpen !== state.open) {
      return { open: nextOpen }
    }
    return null
  }

  /**
   * Is component controlled? (otherwise, it's pure)
   *
   * @returns {boolean}
   */
  isControlled () {
    return this.props.open == null && this.props.info != null
  }

  /**
   * Handle click on marker
   *
   * @param {*} event
   */
  onClick = (event) => {
    const { onClick } = this.props

    // Toggle info window
    this.toggle()

    // Fire external `onClick` event
    if (onClick) {
      onClick(event)
    }
  }

  /**
   * Toggle opening status when component is controlled
   */
  toggle () {
    if (this.isControlled()) {
      this.setState({ open: !this.state.open })
    }
  }

  /**
   * Handle event on closing info window
   */
  onClose = () => {
    const { onClose } = this.props

    // Close info window
    this.toggle()

    // Fire external `onClose` event
    if (onClose) {
      onClose()
    }
  }

  /**
   * Render marker
   *
   * @returns {React.Element}
   */
  render () {
    const { position, info, open } = this.props

    // Check if info window is opened
    const isOpen = this.isControlled() ? this.state.open : info && open

    // Build info window
    const infoWindow = isOpen ? (
      <InfoWindow onCloseClick={this.onClose}>
        <div>{info}</div>
      </InfoWindow>
    ) : null

    // Render marker
    return (
      <GoogleMarker position={position} onClick={this.onClick}>
        {infoWindow}
      </GoogleMarker>
    )
  }
}

Marker.displayName = 'Marker'

Marker.propTypes = propTypes

export default Marker
