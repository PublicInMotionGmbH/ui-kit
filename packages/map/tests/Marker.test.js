import React from 'react'
import { shallow } from 'enzyme'

import Marker from '../src/Marker'

describe('<Marker />', () => {
  let previousGoogle
  let routeFn

  beforeEach(() => {
    routeFn = jest.fn()
    previousGoogle = global.google || window.google
    global.google = window.google = {
      maps: {
        Map: function () {
          this.setCenter = jest.fn()
          this.setOptions = jest.fn()
          this.setZoom = jest.fn()
        },
        DirectionsService: function () {
          this.route = routeFn
        },
        DirectionsRenderer: function () {
          this.setDirections = jest.fn()
          this.setMap = jest.fn()
        },
        LatLng: function (lat, lng) {
          this.lat = lat
          this.lng = lng
        },
        Marker: function () {
          this.setPosition = jest.fn()
          this.setMap = jest.fn()
        },
        TravelMode: {
          DRIVING: 'driving'
        },
        DirectionsStatus: {
          OK: 'ok',
          ERROR: null
        },
        event: {
          addListener: jest.fn()
        }
      }
    }
  })

  afterEach(() => {
    global.google = window.google = previousGoogle
  })

  it('renders children correctly', () => {
    const wrapper = shallow(<Marker position={{lat: -27.5598, lng: 151.9507}} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('onClick open InfoWindow', () => {
    const wrapper = shallow(<Marker position={{lat: -27.5598, lng: 151.9507}} info='InfoWindow is open' />)

    wrapper.find('Marker').prop('onClick')()
    wrapper.update()

    expect(wrapper.find('InfoWindow').length).toBeTruthy()
  })

  it('onCloseClick close infoWindow', () => {
    const wrapper = shallow(<Marker position={{lat: -27.5598, lng: 151.9507}} info='InfoWindow is open' />)

    wrapper.find('Marker').prop('onClick')()
    wrapper.update()

    expect(wrapper.find('InfoWindow').length).toBeTruthy()

    wrapper.find('InfoWindow').prop('onCloseClick')()
    wrapper.update()

    expect(wrapper.find('InfoWindow').length).not.toBeTruthy()
  })

  it('info text is passed to infoWindow', () => {
    const wrapper = shallow(<Marker position={{lat: -27.5598, lng: 151.9507}} info='This is InfoWindow' />)

    wrapper.find('Marker').prop('onClick')()
    wrapper.update()
    expect(wrapper.find('InfoWindow').length).toBeTruthy()

    const infoText = wrapper.find('InfoWindow').props().children.props.children

    expect(infoText).toEqual('This is InfoWindow')
  })
})
