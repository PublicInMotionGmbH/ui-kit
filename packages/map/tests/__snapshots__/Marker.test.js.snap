// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Marker /> renders children correctly 1`] = `
<Marker
  onClick={[Function]}
  position={
    Object {
      "lat": -27.5598,
      "lng": 151.9507,
    }
  }
/>
`;
